### PR TITLE
Reduce memory footprint of InodeLockList

### PR DESCRIPTION
**Pull request by cheng-chang**
Although the Entry abstraction in SimpleInodeLockList makes code more readable, but it unnecessarily stores extra representations of Edges, each consumes a long and a String. Since Each FileSystemMaster operation will create at least one InodeLockList, the lower memory consumption the better.

This PR removes the Entry abstraction without sacrificing readability too much.

### CODE REVIEW

**Comment by cheng-chang** [comment]
jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 1]
```diff
@@ -186,67 +197,72 @@ public void downgradeEdgeToInode(Inode inode, LockMode mode) {
...
-   * If the lock list is write locked, returns LockMode.WRITE. Otherwise, returns the given mode.
+   * If mode is read but the lock list is write locked, precondition check will fail.
    *
    * This helps us preserve the invariant that there is never a READ lock following a WRITE lock.
    *
    * @param mode a lock mode
-   * @return the mode
    */
-  private LockMode nextLockMode(LockMode mode) {
+  private void checkMode(LockMode mode) {
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: I think the implicit lock upgrade makes sense, I changed back.

**Comment by cheng-chang** [comment]
@ggezer thanks for comments, PTAL.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 2]
```diff
@@ -32,20 +31,30 @@
...
   /**
-   * The index of the first write lock entry. All locks before this index are read locks, and all
-   * locks after and including this index are write locks. If all locks are read locks,
-   * mFirstWriteLockIndex == mEntries.size().
+   * If last lock in mLocks is an edge lock, then this is the edge.
+   * Otherwise, null.
+   */
+  private Edge mLastEdge;
+  /**
+   * Lock list. The locks always alternate between Inode lock and Edge lock.
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Is inode lock always first?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 3]
```diff
@@ -32,20 +31,30 @@
...
+   */
+  private List<LockResource> mLocks;
+  /**
+   * The index of the first write lock entry.
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: index into `mLocks`?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 4]
```diff
@@ -56,43 +65,43 @@
    */
   public SimpleInodeLockList(InodeLockManager inodeLockManager) {
     mInodeLockManager = inodeLockManager;
-    mEntries = new ArrayList<>(INITIAL_CAPACITY);
-    mFirstWriteLockIndex = 0;
+    mInodes = new ArrayList<>(INITIAL_CAPACITY);
+    mLocks = new ArrayList<>(INITIAL_CAPACITY);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Is the initial capacity an issue for memory consumption? This change effectively made the initial size twice as large

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 2 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Added comment `The first lock can be either Inode or Edge lock.`

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 5]
```diff
@@ -32,20 +31,30 @@
 public class SimpleInodeLockList implements InodeLockList {
   private static final Logger LOG = LoggerFactory.getLogger(SimpleInodeLockList.class);
 
+  private static final int NO_WRITE_LOCK = -1;
```
[the remark this answers is not in the document]
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Added comment `Default value for {@link #mFirstWriteLockIndex} when there is no write lock.`.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 3 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Updated comment: ` The index of the first write lock entry in {@link #mLocks}`.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 4 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: I changed to reserve less capacity.

**Comment by cheng-chang** [comment]
@gpang PTAL

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 4 above)
Comment (ggezer) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: 2 is too few as it only covers locking a hierarchy of 1 level deep.
Since this struct grows progressively with every lock and we remove only from tail, I believe LinkedList could be of use here instead of ArrayList. WDYT?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 6]
```diff
@@ -56,43 +79,43 @@
...
     mode = nextLockMode(mode);
-    if (!mEntries.isEmpty()) {
+    if (!mLocks.isEmpty()) {
       Preconditions.checkState(!endsInInode(),
           "Cannot lock inode %s for lock list %s because the lock list already ends in an inode",
           inode.getId(), this);
-      String lastEdgeName = ((EdgeEntry) lastEntry()).getEdge().getName();
-      Preconditions.checkState(inode.getName().equals(lastEdgeName),
-          "Expected to lock inode %s but locked inode %s", lastEdgeName, inode.getName());
+      Preconditions.checkState(inode.getName().equals(mLastEdge.getName()),
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: `mLastEdge` could be `null`, right?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 7]
```diff
@@ -186,67 +212,71 @@ public void downgradeEdgeToInode(Inode inode, LockMode mode) {
...
-      mFirstWriteLockIndex = mEntries.size();
+  private void removeLastLock() {
+    mLocks.remove(mLocks.size() - 1).close();
+    if (mFirstWriteLockIndex >= mLocks.size()) {
+      mFirstWriteLockIndex = NO_WRITE_LOCK_INDEX;
+    }
+    if (mLastEdge != null) {
+      mLastEdge = null;
+    } else {
+      Inode last = mInodes.remove(mInodes.size() - 1);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Why is this required? Don't we already unlock and remove earlier?

Could this fail because there is nothing to remove?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 8]
```diff
@@ -186,67 +212,71 @@ public void downgradeEdgeToInode(Inode inode, LockMode mode) {
...
+      if (!mLocks.isEmpty()) {
+        if (mInodes.isEmpty()) {
+          mLastEdge = ROOT_EDGE;
+        } else {
+          mLastEdge = new Edge(mInodes.get(mInodes.size() - 1).getId(), last.getName());
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: Is `mInodes.get(mInodes.size() - 1)` going to be the one before `last`? Do all of these indexes exist?

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 6 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: no, the previous precondition has checked that it ends in an edge, so mLastEdge must not be null.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 7 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: in line https://github.com/Alluxio/alluxio/pull/9092/files#diff-74f13917d59a65b6dcbc4f09e382edd3R265, we only remove the lock, here, we remove the corresponding inode.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 8 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: when this `else` block is executed, neither `mLocks` nor `mInodes`is empty.

**Comment by cheng-chang** [comment]
@gpang PTAL

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 4 above)
Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: My only concern of using LinkedList is that the implementation of `get` is O(n). But I checked that the only usage of `get` is to get the last or the second last element, which is O(1) in LinkedList. So I changed to LinkedList.

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java [hunk 9]
```diff
@@ -257,32 +278,22 @@ public LockMode getLockMode() {
...
-        .map(e -> ((InodeEntry) e).getInode())
-        .collect(Collectors.toList());
+    return new ArrayList<>(mInodes);
   }
 
   @Override
   public Inode get(int index) {
-    int entryIndex = mEntries.get(0) instanceof InodeEntry ? index * 2 : index * 2 + 1;
-    return ((InodeEntry) mEntries.get(entryIndex)).getInode();
+    return mInodes.get(index);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: This will be slow with the linked list now?

I'd imagine most of the usage only wants the last index? Is that efficient for the implementation of the linked list?

[later] Maybe we should consider adding a `getLastInode` api.

Comment (cheng-chang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: LinkedList is a double linked list, so getting the last index is O(1)

**Comment by cheng-chang** [comment]
@gpang PTAL

File: core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java (hunk 9 above)
Comment (gpang) [reply core/server/master/src/main/java/alluxio/master/file/meta/SimpleInodeLockList.java]: looks like LinkedList.get() starts from the end that the index is closer to.

**Comment by cheng-chang** [comment]
alluxio-bot, merge this please
